DakshChan/IBS
# [IBS-12] Add 3 extra features to Aggregated Grades Table

## Overview
- For each task column displayed on "View Grades" page's table (can be clicked on course card as instructor on main page), display the weight of the task relative to course grade next to task name. e.g. instead of `A1` as column name, display `A1 (15%)` to show that the task `A1` is worth 15% of the course grade.
- For each student row on the grades table, display the current overall grade for said student at the last column called `Current Grade`.
- Create backend endpoint for instructor called `/course/:courseId:/get` which retrieves course details.
- Modify heading of the page to display `All Grades of ____` instead of `All Grades for Course ID: __`.
- Implement search functionality for table to search by student names.

### Example of the features in action

#### Default Layout

![Default Layout](https://github.com/DakshChan/IBS/assets/47423409/4ad4c88a-200e-408c-8a5f-394bc69b3ca8)

#### Search bar (autocomplete dropdown)

![Autocomplete Dropdown](https://github.com/DakshChan/IBS/assets/47423409/c1a4d92b-5fc3-4497-8f8b-04c93814320d)

#### Dynamic table results for search

![Dynamic Autocomplete](https://github.com/DakshChan/IBS/assets/47423409/202a91d6-3394-4079-9457-ef63b47d474f)

## Steps for testing (from previous PR -- same as before)
0. Use `make build` and `make run` to build and run the docker containers. On Postman, make sure `base_url` variable is set to `http://localhost/api`. Make sure you have at least one user in DB of role `admin` and `instructor`.
1. From `http://localhost` on web browser, login as `admin` user.
    Ensure at least one course has been added to the database. Otherwise, add one right now.
   1. Create and and new students to the database through `Add role` card on `http://localhost/admin/course/1/task`:
       ```
       Username: student1
       Role: student
       Email: student1@mail.com
       Update user info: false
       ```
   2. Go to `user_info` table on database, and change the password hash values to the hash of `pass`.
   3. From Postman, login as `Instructor`. Create a new task (you can create more than 1) with the following values:
       ```
       task: task1
       due_date: 2023-06-06 12:30:45
       hidden: false
       min_member: 1
       max_member: 3
       max_token: 2
       change_group: true
       hide_interview: false
       ```
   4. Through `POST Add criteria` endpoint on Postman, send a request to backend with the following values:
       ```
       task: task1
       criteria: criteria1
       total: 50
       description: anything here
       ```
   5. To submit a mark for a student, login as an instructor on Postman, and send using the following values to `POST Submit mark` endpoint:
       ```
       task: task1
       criteria: criteria1
       username: student1
       mark: 40
       overwrite: false
       ```
   6. Logged in as an instructor on frontend, head to the following URL: `http://localhost/instructor/all-grades/:courseId`. e.g. to `http://localhost/instructor/all-grades/1`.
   7. You should see a table of all students enrolled in the course, their respective assignments as columns and the students' marks as a table cell in the respective assignment columns.
